### DISCUSSION TITLE
Does Keras require images to be loaded with 3 channels to train data?

### DISCUSSION THREAD

**Discussion by mathieuboudreau** [neuropoly/axondeepseg#244] [Bug reports, suggestions for code refactoring]
Question for you @vs74,

I just fell on a funny bug in one of my branches (see PR #243 ) while refactoring one of your recently modifications to use Keras.

I noticed that you're loading training data with CV2, e.g.

https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/data_management/input_data.py#L39

and that this loads data with three channels (even if the images are in B&W). For masks you don't load with three channels:

https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/data_management/input_data.py#L42

But you create three channels right after anyways to hold your three tissue times:

https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/data_management/input_data.py#L43

https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/data_management/input_data.py#L94-L99

**This feels like it might hog the memory, as it stores 3x the amount of data necessary as before**, and might impact us when training large datasets.

If I try loading these with imageio's imread as 8bit (as we do everywhere else in AxonDeepSeg), Keras throws me an error, as it appears it expects to have more dimensions (because imageio is only giving me 2 dimensions, and CV2 is giving us 3).

```
E           ValueError: ('Input data in `NumpyArrayIterator` should have rank 4. You passed an array with shape', (2, 256, 256))
```

See Travis log: https://travis-ci.org/neuropoly/axondeepseg/builds/566773230?utm_source=github_status&utm_medium=notification

Is there are way to get around this that we can load 8bit B&W images instead of your RGB to get this working with Keras?

**Comment by mathieuboudreau** [comment]
If there's no way around that with Keras, I'm wondering if we could instead load it as greyscale but add a singleton dimension (instead of 3!), using something like this: https://docs.scipy.org/doc/numpy/reference/generated/numpy.expand_dims.html

**Comment by vasudev-sharma** [comment]
> Question for you @vs74,
> 
> I just fell on a funny bug in one of my branches (see PR #243 ) while refactoring one of your recently modifications to use Keras.
> 
> I noticed that you're loading training data with CV2, e.g.
> 
> https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/data_management/input_data.py#L39
> 
> and that this loads data with three channels (even if the images are in B&W). For masks you don't load with three channels:
> 
> https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/data_management/input_data.py#L42
> 
> But you create three channels right after anyways to hold your three tissue times:
> 
> https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/data_management/input_data.py#L43
> 
> https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/data_management/input_data.py#L94-L99
> 
> **This feels like it might hog the memory, as it stores 3x the amount of data necessary as before**, and might impact us when training large datasets.
> 
> If I try loading these with imageio's imread as 8bit (as we do everywhere else in AxonDeepSeg), Keras throws me an error, as it appears it expects to have more dimensions (because imageio is only giving me 2 dimensions, and CV2 is giving us 3).
> 
> ```
> E           ValueError: ('Input data in `NumpyArrayIterator` should have rank 4. You passed an array with shape', (2, 256, 256))
> ```
> 
> See Travis log: https://travis-ci.org/neuropoly/axondeepseg/builds/566773230?utm_source=github_status&utm_medium=notification
> 
> Is there are way to get around this that we can load 8bit B&W images instead of your RGB to get this working with Keras?

Actually the model created under `AxonDeepSeg/network_construction.py` is taking 3 dimension as input image. However I can retrain the model and fix this problem ( further which the image would be of single channel). Just to be on the same page the image loaded after this would be of `cv2.IMREAD_grayscale` type just as the same way mask is loaded using `CV2`. Is this fine ?

**Comment by mathieuboudreau** [comment]
@vs74 Thanks for that - I'll test that out in my branch. Ideally - I'd prefer it if we used imagio instead of cv2 (since all the other I/O handling in ADS uses imageio). Do you think that'll be an issue? Is there a reason why you preferred cv2 to load images, or it was just arbitrary?

**Comment by vasudev-sharma** [comment]
Usage of `CV2` was arbitrary, I am not sure whether `imageio` will lead to an issue ( ideally it shouldn't). While checking this on your branch, could you test out with `imageio` as well?

**Comment by mathieuboudreau** [comment]
Great- thanks! I found out about this feature/issue in a branch where I was trying to consolidate all I/O operations with a custom axondeepseg function to call for consistency, so I'll test it out on that branch too.

https://github.com/neuropoly/axondeepseg/blob/b85496541f38e748601998d8e0d4375a1efb40d2/AxonDeepSeg/ads_utils.py#L253-L271
